### PR TITLE
<input>.value improvements

### LINKED ISSUE #2063: <input type=file>'s value setter is faulty
### Basic info:

- **Node.js version:** v9
- **jsdom version:** master

### Minimal reproduction case

```js
const { JSDOM } = require("jsdom");

const options = {
  runScripts: "dangerously"
};
const dom = new JSDOM(`
  <input id=inp type=file>
  <script>document.getElementById("inp").value = "";</script>
`, options);
```

```
TypeError: 'set' on proxy: trap returned falsish for property 'length'
    at HTMLInputElementImpl.set value [as value] (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/living/nodes/HTMLInputElement-impl.js:195:27)
    at HTMLInputElement.set [as value] (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/living/generated/HTMLInputElement.js:893:25)
    at about:blank:1:38
    at ContextifyScript.Script.runInContext (vm.js:59:29)
    at Object.runInContext (vm.js:133:6)
    at processJavaScript (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/living/nodes/HTMLScriptElement-impl.js:140:10)
    at HTMLScriptElementImpl._eval (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/living/nodes/HTMLScriptElement-impl.js:65:7)
    at /home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/browser/resource-loader.js:31:22
    at Object.check (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/living/nodes/Document-impl.js:91:11)
    at /home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/living/nodes/Document-impl.js:110:12
    at HTMLScriptElementImpl._attach (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/living/nodes/HTMLScriptElement-impl.js:42:72)
    at HTMLBodyElementImpl._attach (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/living/nodes/Node-impl.js:290:15)
    at HTMLBodyElementImpl._attach (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/living/nodes/Element-impl.js:97:11)
    at HTMLHtmlElementImpl._attach (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/living/nodes/Node-impl.js:290:15)
    at HTMLHtmlElementImpl._attach (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/living/nodes/Element-impl.js:97:11)
    at DocumentImpl.insertBefore (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/living/nodes/Node-impl.js:230:22)
    at DocumentImpl.appendChild (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/living/nodes/Node-impl.js:328:17)
    at DocumentImpl.appendChild (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/living/nodes/Document-impl.js:401:18)
    at setChild (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/browser/htmltodom.js:240:18)
    at HTMLToDOM._parseWithParse5 (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/browser/htmltodom.js:47:7)
    at HTMLToDOM._doParse (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/browser/htmltodom.js:28:42)
    at HTMLToDOM.appendToDocument (/home/timothy-gu/dev/jsdom/jsdom/lib/jsdom/browser/htmltodom.js:24:17)
    at new JSDOM (/home/timothy-gu/dev/jsdom/jsdom/lib/api.js:68:29)
```

### How does similar code behave in browsers?

[Empties the element's `files` list.](https://html.spec.whatwg.org/multipage/input.html#dom-input-value-filename)

Regression from e7b68fa33db497aa1ddb5dbfc14f2bc53242e3a8 (which was committed w/o tests).

### PR DESCRIPTION
**Pull request by TimothyGu**
Fixes #2063.

### CODE REVIEW

File: lib/jsdom/living/helpers/dates-and-times.js [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+  }
+  return { year, month, day: undefined };
+}
+
+// https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-month-string
+function isValidMonthString(str) {
+  return parseMonthString(str) !== undefined;
+}
+function serializeMonth({ year, month }) {
+  const yearStr = leftPad(`${year}`, 4, "0");
```
Comment (Sebmaster) [inline lib/jsdom/living/helpers/dates-and-times.js]: String#padStart?

Comment (TimothyGu) [reply lib/jsdom/living/helpers/dates-and-times.js]: We support Node.js versions that don’t have it yet

Comment (Sebmaster) [reply lib/jsdom/living/helpers/dates-and-times.js]: Ah, of course.

**Comment by TimothyGu** [comment]
@Zirro Done. Also incorporated #2065 into the test suite.

**Comment by Sebmaster** [comment]
@TimothyGu Seems like you've got a failure in a number input test there?

**Comment by TimothyGu** [comment]
@Sebmaster It seems to either be an HTML spec bug or a WPT issue. Will file tickets for more details on it.

**Comment by Sebmaster** [comment]
@TimothyGu Should we just ignore the test for now? It's an additional one right, no regression?

**Comment by domenic** [comment]
I think we should make jsdom go against spec for now, matching the test, to get increased test coverage, with a TODO comment to fix when the spec/tests get themselves straightened out.

**Comment by TimothyGu** [comment]
CI fixed.

File: lib/jsdom/living/helpers/dates-and-times.js [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+  }
+  return { year, month };
+}
+
+// https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-month-string
+function isValidMonthString(str) {
+  return parseMonthString(str) !== null;
+}
+function serializeMonth({ year, month }) {
+  const yearStr = leftPad(`${year}`, 4, "0");
```
Comment (Sebmaster) [inline lib/jsdom/living/helpers/dates-and-times.js]: We should drop the templating for these.

Comment (TimothyGu) [reply lib/jsdom/living/helpers/dates-and-times.js]: This is needed to cast `year` to a string.

**Pull request merged**